plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java

```diff
@@ -527,32 +529,43 @@ private Optional<String> getQualifiedParentName(MongoColumnHandle columnHandle)
      * Creates a set of sufficient columns for the input projected columns. For example,
      * if input {@param columns} include columns "a.b" and "a.b.c", then they will be projected from a single column "a.b".
      */
-    public static List<MongoColumnHandle> getSufficientColumns(List<MongoColumnHandle> columnHandles)
+    public static List<MongoColumnHandle> projectSufficientColumns(List<MongoColumnHandle> columnHandles)
     {
         List<MongoColumnHandle> sufficientColumns = new ArrayList<>();
         for (MongoColumnHandle column : columnHandles) {
-            String qualifiedName = column.getQualifiedName();
-            List<MongoColumnHandle> columnsToRemove = new ArrayList<>();
-            boolean isAddable = true;
-
-            for (MongoColumnHandle existingColumn : sufficientColumns) {
-                String name = existingColumn.getQualifiedName();
-                if (name.startsWith(qualifiedName + ".")) {
-                    columnsToRemove.add(existingColumn);
-                }
-                else if (qualifiedName.startsWith(name + ".")) {
-                    isAddable = false;
+            boolean addColumn = true;
+
+            for (int i = sufficientColumns.size() - 1; i >= 0; i--) {
+                MongoColumnHandle existingColumn = sufficientColumns.get(i);
+                if (isParentColumn(existingColumn, column)) {
+                    // If the parent of the visiting column already exists, mark it to indicate no addition is needed
+                    addColumn = false;
                     break;
                 }
+                else if (isParentColumn(column, existingColumn)) {
```

Comment:
if we sort the input column handles based on the size of the dereference column - then we could apply this check once right ?